### PR TITLE
Support service discovery list backup. (#751)

### PR DESCRIPTION
**Pull request by chenzhiguo**
Cache the service list data, if the Consul cluster is not available, use cache data to ensure that the service is highly available.

### CODE REVIEW

**Comment by spencergibb** [comment]
Can you describe the use case here? Both spring cloud loadbalancer and ribbon have caches for loadbalancing. If we decide to do this we should be consistent with s-c-lb https://github.com/spring-cloud/spring-cloud-commons/tree/main/spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/cache

**Comment by chenzhiguo** [comment]
@spencergibb I found that the cache in the load balancer will be lost after a certain period of time. This is not the same as the purpose of my PR. Of course, we can modify the the load balancer cache to achieve the same purpose. 
I don't quite clear what other help do I do?

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java [hunk 1]
```diff
@@ -73,11 +81,20 @@ private void addInstancesToList(List<ServiceInstance> instances, String serviceI
...
-		for (HealthService service : services.getValue()) {
-			instances.add(new ConsulServiceInstance(service, serviceId));
+		try {
+			Response<List<HealthService>> services = this.client.getHealthServices(serviceId, request);
+
+			for (HealthService service : services.getValue()) {
+				instances.add(new ConsulServiceInstance(service, serviceId));
+			}
+			if (properties.isBackup()) {
+				serviceInstanceBackup.put(serviceId, instances);
```
Comment (ryanjbaxter) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: Shouldn't you purge the "backup" before adding the new instances?  You could end up with stale data in the backup.

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java (hunk 1 above)
Comment (chenzhiguo) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: I think that the same serviceId, after being added, the old instances will be covered by new data, isn't it? @ryanjbaxter

Comment (ryanjbaxter) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: What if the service was removed?

Comment (chenzhiguo) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: I think that when the service is removed, consul can also accept the request. The data inside the Backup will be replaced by a empty list. What should this be reasonable? But if consul does not provide a service normally, it should return the last acquisition of the service instance list, no matter whether it is empty.

Comment (ryanjbaxter) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: Sorry i thought this code was iterating all services and not instances of a given service.

Comment (chenzhiguo) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: Yes, but if your assumption, `instances` will be an empty list when there is no service instance returns, which will also be placed in the map. isn't it?

Comment (ryanjbaxter) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryClient.java:91]: Yes correct, I don't think there is a problem after all

**Comment by chenzhiguo** [comment]
Oh...Is there anything unsuitable?

**Comment by spencergibb** [comment]
If the loadbalancer cache looses information, that should be fixed. I've marked #751 for discussion.

**Comment by spencergibb** [comment]
You can change the TTL of the cache to something very large here https://docs.spring.io/spring-cloud-commons/docs/current/reference/html/#loadbalancer-cache-configuration

Otherwise, if you think there is a bug in spring cloud loadbalancer, open an issue in https://github.com/spring-cloud/spring-cloud-commons/issues

**Pull request closed without merging**